### ISSUE TITLE
DEs probably shouldn't link `/share`

### ISSUE THREAD

**Issue by michaelpj** [NixOS/nixpkgs#47173]
## Issue description
A lot of the DEs (gnome, lumina, lqxt, mate, plasma5) put `/share` into `environment.pathsToLink`. While this is a reasonable thing to want, really the responsibility for linking relevant paths should belong with modules associated with those paths. e.g. the SDDM modules should link `/share/sddm`. DEs should only link for the applications they're responsible for (e.g. all the KDE apps in plasma5). 

As it is, if a package only works properly because it's covered by the blanked linking of `/share`, it won't work in a DE that doesn't do this or without a DE. So it masks problems.

However, getting rid of it may be painful as it will likely reveal many such problems.

**Comment by teto** [comment]
As an i3 user, I support the initiative. How to proceed then ?  should we move ` environment.pathsToLink = [ "/share" ];` to the modules. Can't it be a problem for packages without a module ?

**Comment by jtojnar** [comment]
Ideally, packages should not rely on global state. If they do, fix should be attempted first.

**Comment by jtojnar** [comment]
Looking at https://www.freedesktop.org/wiki/Specifications/, we need at least [`sounds`](http://0pointer.de/public/sound-theme-spec.html#sound_lookup) and [`metainfo`](https://www.freedesktop.org/software/appstream/docs/sect-Metadata-Application.html) modules. [`color`](https://www.openicc.org/index.php%3Ftitle=OpenIccDirectoryProposal.html) might be also nice.

**Comment by michaelpj** [comment]
I think packages should only link paths if they expect contributions from other packages at runtime, otherwise they shouldn't. It would probably help to make the description of `pathsToLink` to be a bit more prescriptive about what you should and shouldn't link.

> should we move environment.pathsToLink = [ "/share" ]; to the modules

We should be moving `environment.pathsToLink = [ "/share/my/specific/subfolder" ];` to the modules where possible, yes.

We should probably have a more general `AppStream` module than just `metainfo`, there is also stuff that goes in `app-info` and some other places.

I think we also want a gtk module, and maybe something for gsettings-schemas?

**Comment by jtojnar** [comment]
> I think packages should only link paths if they expect contributions from other packages at runtime, otherwise they shouldn't.

Indeed – packages should be able to load their own data from the prefix that was set during compilation. Only packages that need to be loaded by many other apps should be linked to system path. It might be fine to use global plug-in directory for packages that support plug-ins **temporarily**, but eventually, they should be ported to a wrapper with `symlinkJoin` like e.g. deadbeef.

> We should probably have a more general `AppStream` module than just `metainfo`, there is also stuff that goes in `app-info` and some other places.

Yes, the AppStream specification has a legacy path that would need to be supported as well.

> I think we also want a gtk module, and maybe something for gsettings-schemas?

Those are not a Freedesktop standard but, you are right, we might want to support them. Though on the latter, I keep alternating between accepting that schemas were designed to be global data, and my hate of global state and wanting to eliminate it.

Also, this issue seems to be missing a link to the original freedesktop module: https://github.com/NixOS/nixpkgs/pull/45058

**Comment by michaelpj** [comment]
Yeah, I think we should have modules for stuff that *is* global state, regardless of whether we like it or not. At the moment we're just hackily handling it elsewhere.

One thing I don't like about this whole situation is that it's done with nixos modules so we end up with differences to what gets installed if you use `nix-env`. But that's another kettle of fish.

**Comment by michaelpj** [comment]
I've become much more ambivalent about this. Wanting additional directories in `/share` is something that you want when you install a program that needs them. At the moment, every such program needs to have its own nixos module.

Given that `/share` is basically a bin of implicit system dependencies, maybe we should just link the whole thing by default. This is basically what we do for user profiles in `nix-env`, and I've never heard anyone complain.

**Comment by jtojnar** [comment]
My main point is that if applications rely on their data being linked into a profile, they may subtly break when the app is not installed – for example, when running the built expression from `/nix/store`.

**Comment by michaelpj** [comment]
Hm, that does seem like a good reason to at least wrap the program with a local extension of `XDG_DATA_DIRS`. But when it *is* installed in a profile, then it's annoying that it won't pick up other data files unless we write a module for it.

**Comment by jtojnar** [comment]
Well, if it expects data from other packages, it should accept them as argument, see, for example, https://github.com/NixOS/nixpkgs/blob/e7ff7b93180b5dc3669c8c862fd78d490115a7f0/pkgs/applications/graphics/gimp/wrapper.nix

**Comment by jtojnar** [comment]
For GNOME:

- `share/gnome-control-center/keybindings`
-

**Comment by ryjelsum** [comment]
I have a suggestion that may contribute towards solving this issue, but I want to know if there are any potential issues before I make the pull request:

It seems to me that the `/share/xsessions` and/or `/share/wayland-sessions` paths should be included in pathsToLink if a graphical session is enabled. Fortunately, [we have a module that sets defaults and other ancillary things up, should a display manager or xserver be enabled in the user's configuration.](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/misc/graphical-desktop.nix)

Would putting `environment.pathsToLink = [ "/share/xsessions" "/share/wayland-sessions" ]` inside of that as one of the default options be a productive thing to do to this end? Would it cause any unforeseen problems that I am not anticipating? At the very least, it would make it unnecessary for DEs/WMs themselves to link in these paths themselves, so I think this would be a good place to start.

**Comment by eyJhb** [comment]
> I have a suggestion that may contribute towards solving this issue, but I want to know if there are any potential issues before I make the pull request:
> 
> It seems to me that the `/share/xsessions` and/or `/share/wayland-sessions` paths should be included in pathsToLink if a graphical session is enabled. Fortunately, [we have a module that sets defaults and other ancillary things up, should a display manager or xserver be enabled in the user's configuration.](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/misc/graphical-desktop.nix)
> 
> Would it be productive to put `environment.pathsToLink = [ "/share/xsessions" "/share/wayland-sessions" ]` inside of that as one of the default options to this end? Would it cause any unforeseen problems that I am not anticipating? At the very least, it would make it unnecessary for DEs/WMs themselves to link in these paths themselves, so I think this would be a good place to start.

I made a PR for it now, as I just stumbled upon this issue as well. I'm unsure if this causes any issues, but I don't think it should? https://github.com/NixOS/nixpkgs/pull/411518
